(PUP-9570) Short-circuit compilation if the agent's current env doesn't match the server specified env

joshcooper opened the pull request:
The agent now sends a `check_environment` query parameter when requesting its catalog via the v3 endpoint. The server-side indirector code converts the query parameter to an indirector request option and passes that to the compiler catalog terminus. If the option is `true`, then the server will check if the agent's current environment, that it last pluginsynced in, matches the server-specified environment, as determined by the node terminus, such as an ENC or PE Classifier. If the environments are mismatched, then the terminus short-circuits catalog compilation, returning an empty catalog with the server-specified environment. The agent will detect the mismatch, switch its current environment, and retry its convergence loop.

Prior to this change, if manifests in the server-specified environment referenced a fact that didn't exist in the agent's current environment, then compilation failed and the server returned HTTP 500, causing the agent run to fail.

Note the agent doesn't cache the empty catalog, because that step occurs [after the environment has converged](https://github.com/puppetlabs/puppet/blob/30ade2bb899482cfa52592a0e4ba74f169613579/lib/puppet/configurer.rb#L412-L423).

New agents talking to old servers will behave as before, since old servers ignore the unknown `check_environment` parameter. Similarly, old agents talking to new servers behave as before, because the agent doesn't send the parameter.

GabrielNagy commented on lib/puppet/http/service/compiler.rb:
```diff
@@ -69,6 +69,9 @@ def get_node(name, environment:, configured_environment: nil, transaction_uuid:
   # @param [String] environment The name of the environment we are operating in
   # @param [String] configured_environment Optional, the name of the configured
   #   environment. If unset, `environment` is used.
+  # @param [Boolean] if true, request that the server check if our `environment`
+  #   matches the server specified environment. If they are mismatched, it will
+  #   return an empty catalog with the server specified environment.
```
Should probably move this `@param` after `checksum_type` and also name it

joshcooper replied:
I moved the `check_environment` after the `configured_environment`, and updated the docs to match the same order.

GabrielNagy commented:
LGTM, though I'm not very familiar with this part of the code. Should this also be updated? https://github.com/puppetlabs/puppet/blob/40191235b18136b074514e7bb0d2fc6d2a7058fd/lib/puppet/indirector/catalog/rest.rb#L18-L27

joshcooper commented on lib/puppet/indirector/catalog/compiler.rb:
```diff
@@ -53,6 +53,14 @@ def find(request)
 
     if node.environment
+      # If requested environment doesn't match the server specified environment,
+      # determined by the node temrinus and the requestor wants us to check the
```
typo `temrinus`

joshcooper replied:
updated

joshcooper commented:
Previously I had made `check_environment` default to `true`, but that could be surprising for external consumers of the http client. So I changed the default to `false` and updated the configurer to pass `true` through the rest terminus to the http client.

GabrielNagy commented on spec/integration/application/agent_spec.rb:
```diff
@@ -633,4 +633,35 @@
+
+      server.start_server(mounts: {catalog: catalog_handler}) do |port|
+        Puppet[:serverport] = port
+        expect {
+          agent.command_line.args << '--test' << '--debug'
+          agent.run
+        }.to exit_with(2)
+         .and output(a_string_matching(%r{Notice: Local environment: 'production' doesn't match server specified environment 'newenv', restarting agent run with environment 'newenv'})
+         .and matching(%r{Notice: converged})).to_stdout
```
This test would also pass without the new code, is it worth it to have a test that checks if an empty catalog is returned by the compiler?

joshcooper replied:
Good point, I’ll take a look

joshcooper replied:
I updated the test to be more realistic, so it covers the scenario where a fact doesn't exist in the first environment, but does in the second, and is needed for compilation to succeed. But the actual short-circuit logic can't be tested in the integration test, since it's server side logic that runs in puppetserver. So I updated the `spec/unit/indirector/catalog/compiler_spec.rb` to cover the default behavior and the behavior when `check_environment` is set.

joshcooper commented:
jenkins please test this with servertests

joshcooper commented:
I’ll rebase tgis after the environment guard PR goes in

joshcooper commented:
jenkins please test this with servertests

joshcooper commented on lib/puppet/indirector/catalog/rest.rb:
```diff
@@ -20,6 +20,7 @@ def find(request)
       facts: request.options[:facts_for_catalog],
       environment: request.environment.to_s,
       configured_environment: request.options[:configured_environment],
+      check_environment: request.options[:check_environment],
```
Seems not necessary to use double bang here

joshcooper replied:
nvm, from the commit, "The double negation is needed in case `nil` is passed and behaves the same as `false`."

joshcooper commented:
jenkins please test this with servertests

joshcooper commented:
jenkins please test this with servertests

joshcooper commented:
jenkins please test this with servertests

joshcooper commented:
@ciprianbadescu @Dorin-Pleava @Magisus @justinstoller let me know if you'd like to (re)review, and I'll wait on merging until tomorrow

ciprianbadescu commented:
@joshcooper , it's good to merge for us
